src/main/java/io/numaproj/pulsar/consumer/PulsarSource.java

```diff
@@ -7,139 +7,119 @@
 import io.numaproj.numaflow.sourcer.ReadRequest;
 import io.numaproj.numaflow.sourcer.Server;
 import io.numaproj.numaflow.sourcer.Sourcer;
-
-import jakarta.annotation.PreDestroy;
 import lombok.extern.slf4j.Slf4j;
+
 import org.apache.pulsar.client.api.Consumer;
-import org.apache.pulsar.client.api.PulsarClient;
+import org.apache.pulsar.client.api.Messages;
 import org.apache.pulsar.client.api.PulsarClientException;
-
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
 import org.springframework.stereotype.Component;
 
 import javax.annotation.PostConstruct;
-
 import java.nio.charset.StandardCharsets;
-
 import java.time.Instant;
 import java.util.HashMap;
+import java.util.List;
 import java.util.Map;
 import java.util.concurrent.ConcurrentHashMap;
-import java.util.concurrent.TimeUnit;
-import java.util.List;
 
 @Slf4j
 @Component
 @ConditionalOnProperty(prefix = "spring.pulsar.consumer", name = "enabled", havingValue = "true")
 public class PulsarSource extends Sourcer {
 
-    private final Map<String, org.apache.pulsar.client.api.Message<byte[]>> messages = new ConcurrentHashMap<>();
+    // Map tracking received messages (keyed by Pulsar message ID string)
+    private final Map<String, org.apache.pulsar.client.api.Message<byte[]>> messagesToAck = new ConcurrentHashMap<>();
 
     private Server server;
 
     @Autowired
-    private PulsarClient pulsarClient;
-
-    @Autowired
-    private Consumer<byte[]> pulsarConsumer;
+    private PulsarConsumerManager pulsarConsumerManager;
 
     @PostConstruct
     public void startServer() throws Exception {
-
         server = new Server(this);
         server.start();
         server.awaitTermination();
     }
 
     @Override
     public void read(ReadRequest request, OutputObserver observer) {
-        long startTime = System.currentTimeMillis();
-
-        // minimal check: if outstanding messages haven't been acknowledged, do nothing.
-        if (!messages.isEmpty()) {
+        // If there are messages not acknowledged, return
+        if (!messagesToAck.isEmpty()) {
```

Comment:
Please log the messagesToAck using `log.trace`, for debugging purpose. If we have messages not acked, we would like to enable trace level log to inspect the content of the messages.